src/ess/spectroscopy/indirect/io.py

```diff
@@ -0,0 +1,191 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+import h5py
+import numpy
+import scippnexus
+from scipp import DataArray
+
+from ..types import NormWavelengthEvents, NXspeFileName, NXspeFileNames
+
+
+def to_nxspe(events: NormWavelengthEvents, base: NXspeFileName) -> NXspeFileNames:
+    """Take events, which have been binned in incident wavelength and have monitor
+    counts per bin, and output one NXspe file per setting
+
+    Parameters
+    ----------
+    events: scipp.DataArray
+        The events binned in (setting, event_id, incident_wavelength) with
+        at least 'monitor', 'a3', and 'theta' bin coordinates.
+        The events and the bins must also have an 'incident_wavelength' coordinate.
+    base: str | Path
+        The filename base used to produce each NXspe filename.
+
+    Returns
+    -------
+    :
+        The list of filenames containing the NXspe data. If there are N settings in the
+        input DataArray, N filenames are returned. The names are of the form
+            {base}_{i+1:0{ceil(log10(N+1))}d}.nxspe
+    """
+    from pathlib import Path
+
+    from tqdm import tqdm
+
+    dim = 'setting'
+    length = len(str(events.sizes[dim] + 1))
+    files = []
+    if not isinstance(base, Path):
+        base = Path(base)
+    parent = base.parent
+    if not parent.exists():
+        parent.mkdir(parents=True)
+
+    progress = tqdm(range(events.sizes[dim]))
+    for i in progress:
+        ev = events[dim, i]
+        fn = str(base) + '_' + f'{i+1}'.rjust(length, '0') + '.nxspe'
+        files.append(NXspeFileName(fn))
+        _to_one_nxspe(ev, fn, progress)
+    return NXspeFileNames(files)
+
+
+def _make_group(group: h5py.Group) -> scippnexus.Group:
+    return scippnexus.Group(group, definitions=scippnexus.base_definitions())
+
+
+def _lambda_to_ei(incident_wavelength):
+    from scipp.constants import Planck, neutron_mass
+
+    return ((Planck / incident_wavelength) ** 2 / neutron_mass / 2).to(unit='meV')
+
+
+def _ei_ef_to_en(incident_energy, final_energy):
+    return incident_energy - final_energy
```

Comment:
Why do we need this? How does this differ from `energy_transfer` computed in the workflow?

Reply:
They're the same, but the workflow function goes from incident- and final-wavenumber to energy-transfer directly.
https://github.com/scipp/essspectroscopy/blob/a036038872ef08624211e68d95b96189baa44d52/src/ess/spectroscopy/indirect/conservation.py#L119-L123

This is needed to convert the bin coordinates which don't include incident- or final-wavenumber to energy-transfer without needing to first convert energies to wavenumbers and then back again.

I guess one could argue that a second variant should go into `conservation.py` instead of defining it here.

Reply:
I moved the implementation to `conservation.py`

Reply:
Exactly, it's only a coordinate on the events and it's needed for the bins.